### PR TITLE
Fix: prevent user data insert on failed sending email verification

### LINKED ISSUE #461: Prevent user data to be written to database when POST Register User Failed
**Describe the bug**
Create user using POST Register adds data to database when the HTTP response is not 200. This is found when local database is used.

**To Reproduce**
1. Make sure you set FLASK_ENVIRONMENT_CONFIG to local in ".env".
2. Run python run.py from the terminal.
3. Go to the link created by step above.
4. Go to 'POST/register' and add user (that is not already in the database and using wrong pattern).
5. You should receive response 500.
6. Check the database by opening new terminal and run 'sqlite3 local_data.db;'
7. On sqlite3 shell run 'select * from users;'
8. You should see that the user has been added to the record regardless of the process being failed.

**Expected behavior**
The user should not get added to the database as it did not passed validation.

**Screenshots**

- Swagger API with local database - error on POST/Register.

Screen Shot 2020-02-27 at 11 11 36 am

- Sqlite local data added user regardless failed process.
Screen Shot 2020-02-27 at 11 12 57 am

**Desktop (please complete the following information):**
 - OS: Mac OS 
 - Browser: Safari
 - Version (is this referring to Mac OS version??): Mojave 10.14.6 

**Additional context**
Might need some validation prior to allowing user to be added to database that will cancel the process if HTTP returned with other than 200.

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Include a summary of the change and relevant motivation/context. List any dependencies that are required for this change.

Fixes #461 and replaces PR #463 

### Type of Change:

- Code

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
This is tested by:

-  attempting to create user with 'Less secure app access' turned 'OFF' on gmail settings. 

- checking local database to ensure new data is not inserted

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by mtreacy002** [comment]
@isabelcosta, I had to close issue #463 and replace it with this one as I'm trying to follow best practice by only having one commit per PR. Not sure whether it's expected within AnitaB.org or not, but I saw it being mentioned in few PRs. I think we need to state it clearly in the CONTRIBUTING.md.

I've checked everywhere in that document, and only can see this in regard to squash commits:

> * When you are finished with your work, please squash your commits otherwise we will squash them on your PR (this can help us maintain a clear commit history). 

But it didn't seem to be a strict rule saying that only one commit is allowed for each PR.

Can I ask what is the accepted way here? Can I add each commit separately just to show my thoughts process, or should I use commit --ammend to put new commit  into the same commit already in PR?

**Comment by isabelcosta** [comment]
@mtreacy002 I saw some conversations about this convention on Zulip, but i have not got time yet to answer there. Since we have Squash and Merge option, I agree that this should not be required and I also like having multiple commits, doing little commits as you go. So in my opinion, having multiple commits in the PR is fine, but when merging to `develop` it should go as squashed, and that we maintainers can do that. I will go to Zulip and answer this. I guess we do have to be more clear in the CONTRIBUTING.md. I purposely put the "please squash (...) otherwise we squash", to not be strict. We can also discuss this as community.

**Comment by mtreacy002** [comment]
Thanks, @isabelcosta for your understanding. I agree, we need to discuss this as a community and stated clearly in the guideline. I noticed that some other Open Source organizations do state it as part of their requirement (to have one commit per PR) in their Contributing guideline. But not all open source orgs have the same requirements. So, with the community meeting, hopefully we can clear this out.

**Comment by isabelcosta** [comment]
> Thanks, @isabelcosta for your understanding. I agree, we need to discuss this as a community and stated clearly in the guideline. I noticed that some other Open Source organizations do state it as part of their requirement (to have one commit per PR) in their Contributing guideline. But not all open source orgs have the same requirements. So, with the community meeting, hopefully we can clear this out.

yap, good point! true not all communities work the same, in the end, it depends on what works best for us.

File: app/api/resources/user.py
Comment (isabelcosta) [inline app/api/resources/user.py]: Why is this considered bad request (400) and the other error is 5xx? Its really just a doubt of mine 🤔

File: app/api/resources/user.py
Comment (isabelcosta) [inline app/api/resources/user.py]: Can we use f-strings here and in the `print()` below, this is so that we have consistency throughout the project.
I listed this here in the [Coding Standards](https://github.com/anitab-org/mentorship-backend/wiki/Coding-Standards) wiki page

File: app/api/resources/user.py
Comment (isabelcosta) [inline app/api/resources/user.py]: Can you please use a HTTPStatus code, like it is in the HTTPStatus module (mentioned [Coding Standards](https://github.com/anitab-org/mentorship-backend/wiki/Coding-Standards) wiki page)

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@mtreacy002 I left some request changes regarding the coding style. The logic seems to be correct, although I still have to run and test this locally. This is a great contribution 🙌

**Comment by isabelcosta** [comment]
**A note for people who might test this**, important things to test here is that if you are using a Gmail account (has some security constraints) in the environment variables if there is an error sending email verification, the user is not created. If nothing fails when sending email then the user can be created in the database. Also good to know, if nothing fails on the backend side but the user does not receive an email, we have a `POST /resend_email` service that can be used to resend the verification email to the user.

**Comment by robotjellyzone** [comment]
> **A note for people who might test this**, important things to test here is that if you are using a Gmail account (has some security constraints) in the environment variables if there is an error sending email verification, the user is not created. If nothing fails when sending email then the user can be created in the database. Also good to know, if nothing fails on the backend side but the user does not receive an email, we have a `POST /resend_email` service that can be used to resend the verification email to the user.

Hi, @isabelcosta @mtreacy002 yesterday only I tested this. I am using Gmail account, i registered a new user but I didn't get any error while sending email verification, I got the normal link on my mail id and also, I used swagger to register the user so it was working using that as well !! 
I can't show you the screenshot right now as I have made many changes in my backend code already as I was adding some of my own features regarding one issue to which I am assigned.so sorry for that but yes, i was not getting error... i think some other may confirm this also ..

**Comment by annabauza** [comment]
> > **A note for people who might test this**, important things to test here is that if you are using a Gmail account (has some security constraints) in the environment variables if there is an error sending email verification, the user is not created. If nothing fails when sending email then the user can be created in the database. Also good to know, if nothing fails on the backend side but the user does not receive an email, we have a `POST /resend_email` service that can be used to resend the verification email to the user.
> 
> Hi, @isabelcosta @mtreacy002 yesterday only I tested this. I am using Gmail account, i registered a new user but I didn't get any error while sending email verification, I got the normal link on my mail id and also, I used swagger to register the user so it was working using that as well !!
> I can't show you the screenshot right now as I have made many changes in my backend code already as I was adding some of my own features regarding one issue to which I am assigned.so sorry for that but yes, i was not getting error... i think some other may confirm this also ..

@robotjellyzone google is changing secure setting to higher level after while, please make sure you lower security settings.

**Comment by robotjellyzone** [comment]
Hi @annabauza so should i remove that 2 step-auth which is required in gmail ?

**Comment by annabauza** [comment]
> Hi @annabauza so should i remove that 2 step-auth which is required in gmail ?

@robotjellyzone Access for less secure apps setting should be turned on.

**Comment by mtreacy002** [comment]
@robotjellyzone , what @annabauza said is correct. I have also managed to reproduce the error using gmail. I mentioned it in [PR#463](https://github.com/anitab-org/mentorship-backend/pull/463#issuecomment-608671622) that I have closed, on how I managed to do that on gmail. Pls check it out. Maybe then you can also reproduce the error. 😉

**Comment by mtreacy002** [comment]
> @mtreacy002 I left some request changes regarding the coding style. The logic seems to be correct, although I still have to run and test this locally. This is a great contribution 🙌

Thanks, @isabelcosta, I'll work on it today..

**Comment by mtreacy002** [comment]
@isabelcosta. I've made the changes and squash the commits together (my first time doing the squash, let me know if I messed up). So, now we have SMTPAuthenticationError to catch internal server error caused by gmail security settings 

Screen Shot 2020-04-13 at 12 04 17 pm

**Comment by mtreacy002** [comment]
@isabelcosta , as for other SMTPException errors, I just put them together as bad request. This way, we can separate if the issue is caused by wrong email security settings or other things. I've also used the HTTPStatus library both on @users_ns.response and exceptions, only for UserRegiter class (to minimise conflict with Nandini's PR). I also removed the print() altogether coz it's not really useful (we can see everything on the console log anyway). Let me know what you think. 😉

File: app/api/resources/user.py
Comment (mtreacy002) [reply app/api/resources/user.py]: I changed the other one (5xx)  now to SMTPAuthenticationError and used HTTPStatus.UNAUTHORIZED which gives error code 401 on Swagger automatically. This one (SMTPException) stays the same (400) to cover all other 4xx errors. I differentiate the 401 specifically to allow a customise error message that clearly states the email security settings to make developer aware this requirement needs to be met.

**Comment by robotjellyzone** [comment]
Hi @annabauza @mtreacy002 i am already using gmail settings that too on high for less scure apps !!! but i am not getting that error  , see my screenshots given at this  ---> https://github.com/anitab-org/mentorship-backend/issues/461#issuecomment-612663403

File: app/api/resources/user.py
Comment (mtreacy002) [reply app/api/resources/user.py]: I removed the print() altogether as we can see it in the console log even without this print()

File: app/api/resources/user.py
Comment (mtreacy002) [reply app/api/resources/user.py]: I've made the changes now to the UserRegister response code so it is now using HTTPStatus code and import the library on the top of the file. I didn't do this in my previous commit as it was done before issue #519 and it was consistent to how we did it before the requested issue. I changed only on the UserRegister class not the rest of the file so Nandini will have less merge conflict when PR #532 is being merged. 😉

**Comment by mtreacy002** [comment]
> Hi @annabauza @mtreacy002 i am already using gmail settings that too on high !!! but i am getting that error , see my screenshots given at this ---> [#461 (comment)](https://github.com/anitab-org/mentorship-backend/issues/461#issuecomment-612663403)

@robotjellyzone , is this email also the one you are using in your .env default email>

**Comment by mtreacy002** [comment]
> > Hi @annabauza @mtreacy002 i am already using gmail settings that too on high !!! but i am getting that error , see my screenshots given at this ---> [#461 (comment)](https://github.com/anitab-org/mentorship-backend/issues/461#issuecomment-612663403)
> 
> @robotjellyzone , is this email also the one you are using in your .env default email>

You mean you're not or you are getting the error??

**Comment by robotjellyzone** [comment]
> > > Hi @annabauza @mtreacy002 i am already using gmail settings that too on high !!! but i am getting that error , see my screenshots given at this ---> [#461 (comment)](https://github.com/anitab-org/mentorship-backend/issues/461#issuecomment-612663403)
> > 
> > 
> > @robotjellyzone , is this email also the one you are using in your .env default email>
> 
> You mean you're not? or you are??

Yes, i am using this gmail one as my deafult email . I mean, i am not getting that error

**Comment by mtreacy002** [comment]
> > > > Hi @annabauza @mtreacy002 i am already using gmail settings that too on high !!! but i am getting that error , see my screenshots given at this ---> [#461 (comment)](https://github.com/anitab-org/mentorship-backend/issues/461#issuecomment-612663403)
> > > 
> > > 
> > > @robotjellyzone , is this email also the one you are using in your .env default email>
> > 
> > 
> > You mean you're not? or you are??
> 
> Yes, i am using this gmail one as my deafult email . I mean, i am not getting that error

I've just pushed and squashed another change, Can you test again on my latest code change, please?

**Comment by robotjellyzone** [comment]
But i am not getting that erro so, what should i test for @mtreacy002

**Comment by mtreacy002** [comment]
> But i am not getting that erro so, what should i test for @mtreacy002

Just wondering.... By high you mean your gmail 2 steps ver==on and allow access to less secure app==off (and you removed any app password you have there), right? and you don't have that secure app password/token in .env, just your normal gmail password? And you have tried to clear the cache to clear history of your app access to gmail? mmmm... If these all been done, I'm not sure why you can't reproduce the same error... 🤔

**Comment by mtreacy002** [comment]
so, your Google account security settings are similar to this screenshot?

Screen Shot 2020-04-13 at 1 15 24 pm

**Comment by mtreacy002** [comment]
With that email above, I'll get the error as shown on the video [here](https://drive.google.com/open?id=1d1pDow5S64C6xC7v6nMPju7B3UlfG-zC). My latest code was able to caught the error using SMTPAuthenticationError and user didn't get inserted to the database.

**Comment by mtreacy002** [comment]
@isabelcosta, I've somehow managed to get the local error in processing when I changed the default email in .env to mail.com (you noticed in the screenshot that the authentication from default email was successful). This is caught by SMTPException and user didn't get inserted into the database as we'd like it to happen (as no email verification was sent)

Screen Shot 2020-04-13 at 3 35 33 pm

**Comment by mtreacy002** [comment]
And this was when the smtp process using mail.com was successful (maybe coz this one is using real email address... mail.com is inconsistent in this sometimes it allows random email sometimes it doesn't)

Screen Shot 2020-04-13 at 3 50 00 pm

and user is inserted.

Screen Shot 2020-04-13 at 3 51 06 pm

**Comment by robotjellyzone** [comment]
> > But i am not getting that erro so, what should i test for @mtreacy002
> 
> Just wondering.... By high you mean your gmail 2 steps ver==on and allow access to less secure app==off (and you removed any app password you have there), right? and you don't have that secure app password/token in .env, just your normal gmail password? And you have tried to clear the cache to clear history of your app access to gmail? mmmm... If these all been done, I'm not sure why you can't reproduce the same error... 🤔

Hi @mtreacy002 sorry for late reply :). See, my 2 step verification is on and i am using app password in .env file [not using my gmail password]

**Comment by mtreacy002** [comment]
> > > But i am not getting that erro so, what should i test for @mtreacy002
> > 
> > 
> > Just wondering.... By high you mean your gmail 2 steps ver==on and allow access to less secure app==off (and you removed any app password you have there), right? and you don't have that secure app password/token in .env, just your normal gmail password? And you have tried to clear the cache to clear history of your app access to gmail? mmmm... If these all been done, I'm not sure why you can't reproduce the same error... 🤔
> 
> Hi @mtreacy002 sorry for late reply :). See, my 2 step verification is on and i am using app password in .env file [not using my gmail password]

I see.... that's why you can't reproduce the bug... Don't use app password. just use email password and make sure you remove the app password on google security settings as my screenshot [above](https://github.com/anitab-org/mentorship-backend/pull/527#issuecomment-612729927)

**Comment by robotjellyzone** [comment]
But app password actually allows us to sign into accounts from apps on those devices that don't support this 2 step verification,  it's very useful to use this app password rather than using our own password as it may lead to some security issues .... so, according to me its best to use this app password, as you just need to enter it once so you don't need to remember it...... What do you think?

**Comment by mtreacy002** [comment]
> But app password actually allows us to sigin in to accounts from apps on those devices that don't suppport this 2 step verification,. it's very useful to use this app password rather than using our own password as it may lead to some security issues .... so, according to me its best to use this app password, as you just need to enter it once so you don't need to remember it......

I agree... But not everyone of us (developer) will automatically remember/know about these security combo (2 steps authentication == on and app password assigned and is used on the .env). This PR is made to prevent the user from being inserted in the case of developer forgot to authorise this app though the app password. Hope this makes sense.

**Comment by robotjellyzone** [comment]
Also, this mentorship system will be handled by the community members so, it would be great to just have only one password for all types of mails, rather than using the password of any of the members...Also, since, this app password generates random characters so, it will be highly secure than using personal password : this was just a suggestion @isabelcosta @annabauza @mtreacy002  :)

**Comment by mtreacy002** [comment]
Even once we got developer on board to the habit of dealing with security settings of gmail (if they use one), this code change is still useful in case of other exception as occurred [here](https://github.com/anitab-org/mentorship-backend/pull/527#issuecomment-612759811), that is not related to authentication of email

**Comment by robotjellyzone** [comment]
> > But app password actually allows us to sigin in to accounts from apps on those devices that don't suppport this 2 step verification,. it's very useful to use this app password rather than using our own password as it may lead to some security issues .... so, according to me its best to use this app password, as you just need to enter it once so you don't need to remember it......
> 
> I agree... But not everyone of us (developer) will automatically remember/know about these security combo (2 steps authentication == on and app password assigned and is used on the .env). This PR is made to prevent the user from being inserted in the case of developer forgot to authorise this app though the app password. Hope this makes sense.

Yeah, i agree but i think these instructions could be added into the READme file for developers so that they can easily try it out

**Comment by mtreacy002** [comment]
> Yeah, i agree but i think these instructions could be added into the READme file for developers so that they can easily try it out

I agree with you... We need to add this instruction on READme file... But mistake can still happen from those who forgot to read the READme... 😉 😂

**Comment by mtreacy002** [comment]
> Also, this mentorship system will be handled by the community members so, it would be great to just have only one password for all types of mails, rather than using the password of any of the members...Also, since, this app password generates random characters so, it will be highly secure than using personal password : this was just a suggestion @isabelcosta @annabauza @mtreacy002 :)

I don't know if our maintainers will agree on that (create a community email where everyone can use and access for the development purpose). To do that we have to publicly put the default email address and app password on READme file. There's a risk that the email could be used by others (not limited to AnitaB community as this is an open source project) to send scam emails or do irresponsible things on behalf of AnitB.org.. Just saying...

**Comment by robotjellyzone** [comment]
> > Yeah, i agree but i think these instructions could be added into the READme file for developers so that they can easily try it out
> 
> I agree with you... We need to add this instruction on READme file... But mistake can still happen from those who forgot to read the READme... 😉 😂

Yeah, but actually as can be seen that every other new contributor is starting to set up the apps locally only just by reading the README file and if things will be written well then they won't miss it for sure ... :)

**Comment by mtreacy002** [comment]
Unless we can 100% sure that all contributors will do exactly as what the guideline is on READme file, then I think the SMTPAuthentication part of this code is still needed. The other exception (SMTPException) will still be needed regardless to deal with any unforeseen situation as shown above.
But I'll leave it to the maintainers & admin to decide. @isabelcosta and @annabauza, can we have your opinion on this?

**Comment by robotjellyzone** [comment]
> Unless we can 100% sure that all contributors will do exactly as what the guideline is on READme file, then I think the SMTPAuthentication part of this code is still needed. The other exception (SMTPException) will still be needed regardless to deal with any unforeseen situation as shown above.

Hmm.. yeah that i think will be useful , let me try it out also ...

**Comment by mtreacy002** [comment]
> > Unless we can 100% sure that all contributors will do exactly as what the guideline is on READme file, then I think the SMTPAuthentication part of this code is still needed. The other exception (SMTPException) will still be needed regardless to deal with any unforeseen situation as shown above.
> 
> Hmm.. yeah that i think will be useful , let me try it out also ...

👍 . Don't forget to grab the latest code change I've pushed about few hours ago. 😉

**Comment by robotjellyzone** [comment]
Hi @mtreacy002 you were right, i am getting this kind of bug . Yeah, your PR is working fine great ... [sorry, took more time, as i got busy with another stuff] . 
But i still think that app password will be  more reliable :) [it was just a matter of decision taken by admins]

**Comment by mtreacy002** [comment]
@robotjellyzone. Thanks for testing the PR. I agree.. App password will still be needed regardless. This is just for precautionary measure in the event this requirement isn't met.
@isabelcosta, I can see there's issue #480 already to add a Troubleshooting.md with first issue tackling this google email security settings Would it be ok if I try work on it? Is it still available?

**Review by robotjellyzone — DISMISSED** [review]

**Comment by mtreacy002** [comment]
Sorry, @Isabel... just wondering what that circle button do ... 😂 (re-request-review). I gave my responses to your comment in last review...Pls let me know what you think... 😉

**Comment by mtreacy002** [comment]
I only just discover it now (re-request review circle thingy). Maybe that's why @ramitsawhney27 didn't realised that I had posted my responses to his comment on the other [PR#463](https://github.com/anitab-org/mentorship-backend/pull/463) coz I didn't press that button (didn't know/notice it back then) 😂 .

**Comment by mtreacy002** [comment]
mmm...not sure why my 'resolve merge conflict' (last commit) removed your stale review, @robotjellyzone. Feel free to add your review again when you have time 😉.

File: app/api/resources/user.py
Comment (devkapilbansal) [inline app/api/resources/user.py:282]: I dont' think we need this extra variable here. When mail is not send, `except` blocks will return error messages and will not proceed further

File: app/api/resources/user.py
Comment (devkapilbansal) [inline app/api/resources/user.py:294]: This condition will always hold as discussed above.

File: app/messages.py
Comment (devkapilbansal) [inline app/messages.py:289]: Need a new blank line at the end.

File: app/api/resources/user.py
Comment (devkapilbansal) [inline app/api/resources/user.py:19]: Instead of star importing, I would suggest writing names of all functions/modules required.

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@mtreacy002 please look into my request below :point_down:

File: app/api/resources/user.py
Comment (epicadk) [reply app/api/resources/user.py:19]: I agree with kapil here certain lint checks discourage wildcard imports.

**Comment by mtreacy002** [comment]
@devkapilbansal , @epicadk and @isabelcosta . This is a very old issue (last changed was 11 month ago) and I think the bug had been fixed by one of the commits that was merged to the repo (I've just pulled rebase from upstream to make the branch up to date). I tried to reproduce the bug using my gmail account with 2-steps verification on and 3rd party access on (without adding MS app in the list of authorised app) (which successfully produced the bug last time), but this time the bug didn't show up. Can you please confirm if this issue exist on your end? If this issue is not valid anymore, feel free to close the issue 😉

**Pull request closed without merging**
